components/raftstore/src/store/snapshot_backup.rs

```diff
@@ -0,0 +1,345 @@
+        &self,
+        tx: UnboundedSender<CheckAdminResponse>,
+    ) -> crate::Result<()>;
+}
+
+impl<EK: KvEngine, ER: RaftEngine> SnapshotBrHandle for Arc<Mutex<RaftRouter<EK, ER>>> {
+    fn send_wait_apply(&self, region: u64, req: SnapshotBrWaitApplyRequest) -> crate::Result<()> {
+        let msg = SignificantMsg::SnapshotBrWaitApply(req);
+        metrics::SNAP_BR_WAIT_APPLY_EVENT.sent.inc();
+        self.lock().unwrap().significant_send(region, msg)
+    }
+
+    fn broadcast_wait_apply(&self, req: SnapshotBrWaitApplyRequest) -> crate::Result<()> {
+        let msg_gen = || {
+            metrics::SNAP_BR_WAIT_APPLY_EVENT.sent.inc();
+            PeerMsg::SignificantMsg(SignificantMsg::SnapshotBrWaitApply(req.clone()))
+        };
+        self.lock().unwrap().broadcast_normal(msg_gen);
+        Ok(())
+    }
+
+    fn broadcast_check_pending_admin(
+        &self,
+        tx: UnboundedSender<CheckAdminResponse>,
+    ) -> crate::Result<()> {
+        self.lock().unwrap().broadcast_normal(|| {
+            PeerMsg::SignificantMsg(SignificantMsg::CheckPendingAdmin(tx.clone()))
+        });
+        Ok(())
+    }
+}
+
+#[derive(Default)]
+pub struct RejectIngestAndAdmin {
+    until: AtomicU64,
+    initialized: AtomicBool,
+}
+
+impl RejectIngestAndAdmin {
+    pub fn register_to(self: &Arc<Self>, coprocessor_host: &mut CoprocessorHost<impl KvEngine>) {
+        let reg = &mut coprocessor_host.registry;
+        reg.register_query_observer(0, BoxQueryObserver::new(Arc::clone(self)));
+        reg.register_admin_observer(0, BoxAdminObserver::new(Arc::clone(self)));
+        info!("registered reject ingest and admin coprocessor to TiKV.");
+    }
+
+    pub fn remained_secs(&self) -> u64 {
+        self.until
+            .load(Ordering::Acquire)
+            .saturating_sub(epoch_second_coarse())
+    }
+
+    pub fn allowed(&self) -> bool {
+        let mut v = self.until.load(Ordering::Acquire);
+        if v == 0 {
+            return true;
+        }
+        let mut expired = v < epoch_second_coarse();
+        while expired {
+            match self
+                .until
+                .compare_exchange(v, 0, Ordering::SeqCst, Ordering::SeqCst)
+            {
+                Ok(_) => {
+                    metrics::SNAP_BR_SUSPEND_COMMAND_LEASE_UNTIL.set(0);
+                    metrics::SNAP_BR_LEASE_EVENT.expired.inc();
+                    break;
+                }
+                Err(new_val) => {
+                    v = new_val;
+                    expired = v < epoch_second_coarse();
+                }
+            }
+        }
+
+        expired
+    }
+
+    pub fn connected(&self) -> bool {
+        self.initialized.load(Ordering::Acquire)
+    }
+
+    /// Extend the lease.
+    ///
+    /// # Returns
+    ///
+    /// Whether previously there is a lease.
+    pub fn heartbeat(&self, lease: Duration) -> bool {
+        let mut v = self.until.load(Ordering::SeqCst);
+        let now = epoch_second_coarse();
```

Comment:
Instant::now_coarse?

Reply:
We need to store it as an atomic variable, I think it might be more convenient if we store it as an `uint64` instead of an `Instant`.